storage/src/mmr/operation_array.rs

```diff
@@ -0,0 +1,220 @@
+    InvalidValue(<V as Array>::Error),
+}
+
+/// The types of operations that change a key's state in the store.
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub enum Type<V: Array> {
+    /// Indicates the key no longer has a value.
+    Deleted,
+
+    /// Indicates the key now has the wrapped value.
+    Update(V),
+}
+
+/// An `Array` implementation for operations applied to a `MutableMmr` K/V store.
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+#[repr(transparent)]
+pub struct Operation<K: Array, V: Array> {
+    pub data: Vec<u8>,
+    _phantom: std::marker::PhantomData<(K, V)>,
+}
+
+impl<K: Array, V: Array> Operation<K, V> {
+    const DELETE_CONTEXT: u8 = 0;
+    const UPDATE_CONTEXT: u8 = 1;
+
+    /// Create a new operation from `key` and `op_type`.
+    pub fn new(key: K, op_type: Type<V>) -> Self {
+        let mut data = Vec::with_capacity(Self::LEN_ENCODED);
+        data.extend_from_slice(&key);
+        match op_type {
+            Type::Deleted => {
+                data.push(Self::DELETE_CONTEXT);
+                data.extend(vec![0u8; V::LEN_ENCODED]);
+            }
+            Type::Update(value) => {
+                data.push(Self::UPDATE_CONTEXT);
+                data.extend_from_slice(&value);
+            }
+        }
+
+        Self {
+            data,
+            _phantom: std::marker::PhantomData,
+        }
+    }
+
+    pub fn to_key(&self) -> K {
+        K::try_from(&self.data[..K::LEN_ENCODED]).unwrap()
+    }
+
+    pub fn to_type(&self) -> Type<V> {
+        if self.data[K::LEN_ENCODED] == 0 {
+            return Type::Deleted;
+        }
+
+        Type::Update(V::try_from(&self.data[K::LEN_ENCODED + 1..]).unwrap())
+    }
+
+    pub fn to_op(&self) -> V {
+        match self.to_type() {
+            Type::Deleted => panic!("No value for delete operation"),
+            Type::Update(v) => v,
+        }
+    }
+}
+
+impl<K: Array, V: Array> Codec for Operation<K, V> {
+    fn write(&self, writer: &mut impl Writer) {
+        writer.write_fixed(&self.data);
+    }
+
+    fn read(reader: &mut impl Reader) -> Result<Self, CodecError> {
+        let value = reader.read_n_bytes(Self::LEN_ENCODED)?;
+        Self::try_from(value.as_ref()).map_err(|e| CodecError::Wrapped("Operation", e.into()))
+    }
+
+    fn len_encoded(&self) -> usize {
+        Self::LEN_ENCODED
+    }
+}
+
+impl<K: Array, V: Array> SizedCodec for Operation<K, V> {
+    const LEN_ENCODED: usize = K::LEN_ENCODED + 1 + V::LEN_ENCODED;
+}
+
+impl<K: Array, V: Array> Array for Operation<K, V> {
+    type Error = Error<K, V>;
+}
+
+impl<K: Array, V: Array> TryFrom<&[u8]> for Operation<K, V> {
+    type Error = Error<K, V>;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        if value.len() != Self::LEN_ENCODED {
+            return Err(Error::InvalidLength);
+        }
+        let mut data = Vec::with_capacity(Self::LEN_ENCODED);
+        data.extend_from_slice(&value[..K::LEN_ENCODED]);
```

Comment:
I believe we can do do `K/V::read_from` over `value`?

Reply:
I don't follow?

Reply:
I did simplify this implementation by using try_from directly from the slice.  Wasn't able to figure out how read_from helps here?